src/asynchronous/client.rs

```diff
@@ -218,3 +159,153 @@ impl Drop for ClientClose {
         trace!("All client is droped");
     }
 }
+
+#[derive(Debug)]
+struct ClientBuilder {
+    rx: Option<MessageReceiver>,
+    streams: Arc<Mutex<HashMap<u32, ResultSender>>>,
+}
+
+impl Builder for ClientBuilder {
+    type Reader = ClientReader;
+    type Writer = ClientWriter;
+
+    fn build(&mut self) -> (Self::Reader, Self::Writer) {
+        let (notifier, waiter) = shutdown::new();
+        (
+            ClientReader {
+                shutdown_waiter: waiter,
+                streams: self.streams.clone(),
+            },
+            ClientWriter {
+                rx: self.rx.take().unwrap(),
+                shutdown_notifier: notifier,
+
+                streams: self.streams.clone(),
+            },
+        )
+    }
+}
+
+struct ClientWriter {
+    rx: MessageReceiver,
+    shutdown_notifier: shutdown::Notifier,
+
+    streams: Arc<Mutex<HashMap<u32, ResultSender>>>,
+}
+
+#[async_trait]
+impl WriterDelegate for ClientWriter {
+    async fn recv(&mut self) -> Option<GenMessage> {
+        self.rx.recv().await
+    }
+
+    async fn disconnected(&self, msg: &GenMessage, e: Error) {
```

Comment:
disconnected is an adj, a verb like disconnect is better

Reply:
fixed